### ISSUE TITLE
Null exception when saving file with pivot table

### ISSUE THREAD

**Issue by stevenxi** [JanKallman/EPPlus#127]
When open and save a Excel file that contains pivot table, which uses complex formula as data source (etc =Offset...), EPPlus throws null exception when save the file.

I've attached the sample file, also created a fix pull request #126 
[PivotTableTestCase.xlsx](https://github.com/JanKallman/EPPlus/files/1727642/PivotTableTestCase.xlsx)

**Comment by stevenxi** [comment]
Pull #126

**Comment by giuliohome** [comment]
I have a big 69MB xlsm full of connections, pivot tables and conditional formatting. 
At save it throws with this null exception about pivot tables...
I've downloaded the latest package from nuget. Is this pull request already included?
What can I do? Help somehow?
Is there a workaround for the file (a code to spot which of the hundreds  table-objects is causing the issue?)

**Comment by stevenxi** [comment]
@giuliohome  I don't think this pull request been merged yet.
And, this pull request only fixes the bug that when using a formula based data range as source pivot tables.
I think it would be better that you create your own nuget feed from a fork, just like what I did. The work progress of EPPlus is bit slow, so waiting for an official fix would take time.

**Comment by giuliohome** [comment]
Thank you! Why not? It could be the same.
I'll try tomorrow morning. 
Fingers crossed :wink:

**Comment by gaston-e-nusimovich** [comment]
I did a spike to help expose what might be be going on. It is a small console app (see the code sample below).

I did a test run with the example workbook provided by stevenxi, and the stdout of said run is as follows (see stdout below).

The spike shows that the library can properly navigate through the collection of Pivot Tables on any given WorkSheet of the open WorkBook, and within any given Pivot Table, it can properly retrieve the XML for said Pivot Table.

In fact, the spike shows that the library can do most things properly with a workbook that has one or more Pivot Tables, but there seems to be a bug when trying to save said workbook.

The code and the test run clearly show that the exception is thrown when the SaveAs is attempted to be executed, in this case, to no avail.

The use of SaveAs to a different file, instead of Save to the same file, allows to proof that there is no weird stuff going on at runtime between the read and the write, like a schema downgrade that is OK in memory, but turns out incompatible when save back to the same file.

I hope that this is helpful, somehow.

Kind regards, GEN



STDOUT

Instantiation is done!
PivotTable Name: PivotTable1
<?xml version="1.0" encoding="UTF-8" standalone="yes"?><pivotTableDefinition xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" name="PivotTable1
" cacheId="24" applyNumberFormats="0" applyBorderFormats="0" applyFontFormats="0" applyPatternFormats="0" applyAlignmentFormats="0" applyWidthHeightFormats="1"
dataCaption="Values" updatedVersion="6" minRefreshableVersion="3" useAutoFormatting="1" itemPrintTitles="1" createdVersion="6" indent="0" outline="1" outlineDat
a="1" multipleFieldFilters="0"><location ref="A1:F6" firstHeaderRow="1" firstDataRow="3" firstDataCol="1" /><pivotFields count="3"><pivotField axis="axisRow" su
btotalTop="0" showAll="0"><items count="3"><item x="0" /><item x="1" /><item t="default" /></items></pivotField><pivotField axis="axisCol" subtotalTop="0" showA
ll="0"><items count="3"><item x="0" /><item x="1" /><item t="default" /></items></pivotField><pivotField axis="axisCol" subtotalTop="0" showAll="0"><items count
="3"><item x="0" /><item x="1" /><item t="default" /></items></pivotField></pivotFields><rowFields count="1"><field x="0" /></rowFields><rowItems count="3"><i><
x /></i><i><x v="1" /></i><i t="grand"><x /></i></rowItems><colFields count="2"><field x="1" /><field x="2" /></colFields><colItems count="5"><i><x /><x /></i><
i t="default"><x /></i><i><x v="1" /><x v="1" /></i><i t="default"><x v="1" /></i><i t="grand"><x /></i></colItems><pivotTableStyleInfo name="PivotStyleLight16"
 showRowHeaders="1" showColHeaders="1" showRowStripes="0" showColStripes="0" showLastColumn="1" /><extLst><ext uri="{962EF5D1-5CA2-4c93-8EF4-DBF5C05439D2}" xmln
s:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main"><x14:pivotTableDefinition hideValuesRow="1" xmlns:xm="http://schemas.microsoft.com/office/
excel/2006/main" /></ext><ext uri="{747A6164-185A-40DC-8AA5-F01512510D54}" xmlns:xpdl="http://schemas.microsoft.com/office/spreadsheetml/2016/pivotdefaultlayout
"><xpdl:pivotTableDefinition16 /></ext></extLst></pivotTableDefinition>
Recalc is done!
Exception: Error saving file D:\data\SaveAsTest.xlsx
Stack Trace:    at OfficeOpenXml.ExcelPackage.Save()
   at OfficeOpenXml.ExcelPackage.SaveAs(FileInfo file)
   at EpPlusPivotTableConsoleApp.Program.Main(String[] args) in d:\Spikes\EpPlusPivotTableConsoleApp\EpPlusPivotTableConsoleApp\Program.cs:line 45


SPIKE SOURCE CODE:

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Configuration;
using System.IO;
using OfficeOpenXml;
using OfficeOpenXml.Table.PivotTable;
using System.Xml;


namespace EpPlusPivotTableConsoleApp
{
  class Program
  {
    static void Main ( string[] args )
    {
      string sWorkbookFileName = ConfigurationManager.AppSettings["WorbookFileName"];
      string sSaveAsWorbookFileName = ConfigurationManager.AppSettings["SaveAsWorbookFileName"];
      XmlDocument xmlDocument;
      StringWriter stringWriter = new StringWriter();
      XmlTextWriter xmlTextWriter = new XmlTextWriter ( stringWriter );

      try
      {
        FileInfo fiWorkbook = new FileInfo ( sWorkbookFileName );
        FileInfo fiSaveAs = new FileInfo ( sSaveAsWorbookFileName );
        ExcelPackage ep = new ExcelPackage ( fiWorkbook );
        Console.WriteLine ( "Instantiation is done!" );

        foreach ( ExcelWorksheet epSheet in ep.Workbook.Worksheets )
        {
          foreach ( ExcelPivotTable epPivotTable in epSheet.PivotTables )
          {
            Console.WriteLine ( "PivotTable Name: {0}", epPivotTable.Name );
            xmlDocument = epPivotTable.PivotTableXml;
            xmlDocument.WriteTo ( xmlTextWriter );
            Console.WriteLine ( "{0}", stringWriter.ToString() );
          }
        }

        ep.Workbook.Calculate();
        Console.WriteLine ( "Recalc is done!" );
        ep.SaveAs ( fiSaveAs );
        Console.WriteLine ( "Save As is done!" );
      }

      catch ( Exception ex )
      {
        Console.WriteLine ( "Exception: {0}", ex.Message );
        Console.WriteLine ( "Stack Trace: {0}", ex.StackTrace );
      }
    }
  }
}

**Comment by stevenxi** [comment]
Hi @gaston-e-nusimovich,
I don't see your console app relate to my test case.

Please check the pull request #126, I've included the sample sheet "EPPlusTest/Workbooks/PivotTableTestCase.xlsx".  Please use that sheet to 'save as'.

Regards,
-Steven

**Comment by giuliohome** [comment]
Hi @gaston-e-nusimovich  !
maybe did you skip the part of the stack trace containing `SavePivotTables()` ?
The crash comes from there _**also in my case**_, that's why I hope this PR can fix also my issue. 
And yes, I did a `SaveAs`

Thanks again to @stevenxi for his support

**Comment by gaston-e-nusimovich** [comment]
@stevenxi In the test run of my spike, I DID use your workbook, PivotTableTestCase.xlsx.

In my spike, it throws the exception when the code tries to execute the SaveAs, as for the rest of the code, it executes without a hitch.

If you open with workbook file with a zip file reader, like for instance 7-zip, and you get to the XML document for the Pivot Table, you can see that the XML of say document is identical to what is shown by spike, so, the EPPLUS library has NO problems working with an existing workbook that contains Pivot Tables, but the library DOES have a bug when it tries to save a workbook that contains one Pivot Tables.

Kind regards, GEN

**Comment by stevenxi** [comment]
@gaston-e-nusimovich Sorry, didn't see it. As you didn't paste your config file.
But what's the point of your spike?  All this bug is about SaveAs failure...

**Comment by gaston-e-nusimovich** [comment]
I wanted to explore a few things, like for instance, whether or not the libray can operate with Pivot Tables, and the point is that it can do it with no problem, and I also wanted to see if the stack trace could give some idea about the possible root cause for the problem, unfortunately it does not show anything beyond the fact that the problem lies within the Save method.

Once we have settled this with this spike, I can try some other spike trajectories in order to explore possible sources of the problem at work in this bug.

I will keep posted all interested parties on this bug through this means.

Kind regards, GEN

**Comment by stevenxi** [comment]
@gaston-e-nusimovich reason for this bug is clear.

Within that .xlsx file, the source of Pivot Table is a formula, which doesn't exist by default. 
When saving the file using EPPlus, it try to load the source, which is null.  That's where the bug happens.

As I said above, the new unit test exposes the bug and also includes the fix already.

**Comment by giuliohome** [comment]
Hi @stevenxi 
I'm still convinced that I have to deal with the same bug and I think your fix is working, 
but unfortunately I can't really test it, because I see that there is another difference between the current official production nuget and the development branch: it is likely another pull [Fix FormulaR1C1 Translate Issue](https://github.com/JanKallman/EPPlus/commit/f481e9d3de608b2db844038f69632bb08727b2ae#diff-b1669011af1ca17e1cd553ece7672e7e) (maybe, or something similar) and that is causing a crash in the `SaveTable()` before the `SavePivotTables()` (when I replace the dll from your nuget in my proj)

just fyi
```
System.ArgumentOutOfRangeException: La lunghezza non può essere minore di zero.
Nome parametro: length
   in System.String.Substring(Int32 startIndex, Int32 length)
   in OfficeOpenXml.ExcelCellBase.ToAbs_1(String part, Int32 row, Int32 col, Int32 rowIncr, Int32 colIncr)
   in OfficeOpenXml.ExcelCellBase.ToAbs(String part, Int32 row, Int32 col, Int32 rowIncr, Int32 colIncr)
   in OfficeOpenXml.ExcelCellBase.Translate(String value, dlgTransl addressTranslator, Int32 row, Int32 col, Int32 rowIncr, Int32 colIncr)
   in OfficeOpenXml.ExcelCellBase.TranslateFromR1C1(String value, Int32 row, Int32 col)
   in OfficeOpenXml.ExcelWorksheet.SaveTables()
   in OfficeOpenXml.ExcelWorksheet.Save()
   in OfficeOpenXml.ExcelWorkbook.Save()
   in OfficeOpenXml.ExcelPackage.Save()
   --- Fine della traccia dello stack dell'eccezione interna ---
   in OfficeOpenXml.ExcelPackage.Save()
```

In conclusion, thank you for your work, anyway I'm afraid that the xlsm file of my business client is too complex for the current status of any Excel library and I need to talk to them so that we can get to a different solution architecture.

**Comment by stevenxi** [comment]
@giuliohome if you can adapt your solution to use .xlsx, probably you can try out the earlier version of EPPlus (

**Comment by JanKallman** [comment]
This is fixed in 4.5.1
